app/src/main/java/com/twilio/video/app/ui/room/RoomActivity.java

```diff
@@ -499,37 +488,26 @@ void connectButtonClick() {
         Editable text = roomEditText.getText();
         if (text != null) {
             final String roomName = text.toString();
-            // obtain latest environment preferences
-            RoomProperties roomProperties =
-                    new RoomProperties.Builder()
-                            .setName(roomName)
-                            .setTopology(
-                                    Topology.fromString(
-                                            sharedPreferences.getString(
-                                                    Preferences.TOPOLOGY,
-                                                    Preferences.TOPOLOGY_DEFAULT)))
-                            .setRecordOnParticipantsConnect(
-                                    sharedPreferences.getBoolean(
-                                            Preferences.RECORD_PARTICIPANTS_ON_CONNECT,
-                                            Preferences.RECORD_PARTICIPANTS_ON_CONNECT_DEFAULT))
-                            .createRoomProperties();
-
-            Single<Room> connection =
-                    updateEnv()
-                            .andThen(tokenService.getToken(displayName, roomProperties))
-                            .flatMap(token -> connect(token, roomName));
-
-            connection
-                    .observeOn(AndroidSchedulers.mainThread())
-                    .doFinally(rxDisposables::clear)
-                    .doOnSubscribe(disposable -> InputUtils.hideKeyboard(this))
-                    .subscribe(roomManager.getRoomConnectionObserver());
+
+            roomViewModel.connectToRoom(
```

Comment:
What are the best practices with respect to exceptions that are thrown as a result of calling a `suspend` function? Should there be a `try catch` clause that wraps the invocation or does that only apply inside a co-routine? If I look at the implementation of this method I see that we use a try catch block to post the TokenError event.

Reply:
Yeah so one of the great things about coroutines is that they allow you to wrap suspend function calls in a try catch to handle any errors that get thrown while executing a coroutine. It is more readable versus a separate callback that returns an error with libraries like RxJava. If I were to remove the try catch, it would propagate up the stack just like a regular exception getting thrown outside of a coroutine.

Reply:
So to answer your question, what I did here would be the way to handle an exception thrown in a coroutine. The reason why I am emitting an error to the live data property in the room manager, is so that my UI can react to the error and present the correct UI.